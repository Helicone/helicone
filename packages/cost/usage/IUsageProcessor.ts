import { ModelUsage } from "./types";
import { Result } from "../../common/result";

export interface ParseInput {
  responseBody: string;
  requestBody?: string;
  isStream: boolean;
}

export interface IUsageProcessor {
  parse(parseInput: ParseInput): Promise<Result<ModelUsage, string>>;
}