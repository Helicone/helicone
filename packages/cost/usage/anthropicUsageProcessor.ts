import { IUsageProcessor, ParseInput } from "./IUsageProcessor";
import { ModelUsage } from "./types";
import { Result } from "../../common/result";

export class AnthropicUsageProcessor implements IUsageProcessor {
  public async parse(parseInput: ParseInput): Promise<Result<ModelUsage, string>> {
    try {
      if (parseInput.isStream) {
        return this.parseStreamResponse(parseInput.responseBody);
      } else {
        return this.parseNonStreamResponse(parseInput.responseBody);
      }
    } catch (error) {
      return {
        data: null,
        error: `Failed to parse Anthropic usage: ${error}`,
      };
    }
  }

  private parseNonStreamResponse(responseBody: string): Result<ModelUsage, string> {
    try {
      const parsedResponse = JSON.parse(responseBody);
      const usage = this.extractUsageFromResponse(parsedResponse);
      
      return {
        data: usage,
        error: null,
      };
    } catch (error) {
      return {
        data: null,
        error: `Failed to parse non-stream response: ${error}`,
      };
    }
  }

  private parseStreamResponse(responseBody: string): Result<ModelUsage, string> {
    try {
      const lines = responseBody
        .split("\n")
        .filter((line) => line.trim() !== "")
        .map((line) => {
          if (line.startsWith("data: ")) {
            try {
              const dataStr = line.replace("data: ", "").trim();
              if (dataStr === "[DONE]") return null;
              return JSON.parse(dataStr);
            } catch {
              return null;
            }
          }
          return null;
        })
        .filter((data) => data !== null);

      const consolidatedUsage = this.consolidateStreamUsage(lines);
      const usage = this.extractUsageFromResponse(consolidatedUsage);

      return {
        data: usage,
        error: null,
      };
    } catch (error) {
      return {
        data: null,
        error: `Failed to parse stream response: ${error}`,
      };
    }
  }

  private consolidateStreamUsage(streamData: any[]): any {
    // message_start and message_delta events that contain usage
    let finalUsage = null;
    
    for (const chunk of streamData) {
      if (chunk?.type === "message_start" && chunk?.message?.usage) {
        finalUsage = chunk.message.usage;
      }
      if (chunk?.type === "message_delta" && chunk?.usage) {
        // message_delta contains the final usage, merge with any existing usage
        finalUsage = { ...finalUsage || {}, ...chunk.usage };
      }
    }

    return { usage: finalUsage };
  }

  protected extractUsageFromResponse(parsedResponse: any): ModelUsage {
    if (!parsedResponse || typeof parsedResponse !== "object") {
      return {
        input: 0,
        output: 0,
      };
    }

    const usage = parsedResponse.usage || {};
    
    const inputTokens = usage.input_tokens ?? 0;
    const outputTokens = usage.output_tokens ?? 0;
    const cacheReadInputTokens = usage.cache_read_input_tokens ?? 0;
    
    const cacheCreation = usage.cache_creation || {};
    const ephemeral5mTokens = cacheCreation.ephemeral_5m_input_tokens ?? 0;
    const ephemeral1hTokens = cacheCreation.ephemeral_1h_input_tokens ?? 0;
    
    const modelUsage: ModelUsage = {
      input: inputTokens,
      output: outputTokens,
    };

    if (cacheReadInputTokens > 0 || ephemeral5mTokens > 0 || ephemeral1hTokens > 0) {
      modelUsage.cacheDetails = {
        cachedInput: cacheReadInputTokens,
      };
      
      if (ephemeral5mTokens > 0) {
        modelUsage.cacheDetails.write5m = ephemeral5mTokens;
      }
      
      if (ephemeral1hTokens > 0) {
        modelUsage.cacheDetails.write1h = ephemeral1hTokens;
      }
    }

    return modelUsage;
  }
}
