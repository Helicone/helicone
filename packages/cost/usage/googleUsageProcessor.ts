import { IUsageProcessor, ParseInput } from "./IUsageProcessor";
import { ModelUsage } from "./types";
import { Result } from "../../common/result";

export class GoogleUsageProcessor implements IUsageProcessor {
  public async parse(parseInput: ParseInput): Promise<Result<ModelUsage, string>> {
    try {
      const response = JSON.parse(parseInput.responseBody);

      // Google AI Studio (Gemini) format
      if (response.usageMetadata) {
        return {
          data: {
            input: response.usageMetadata.promptTokens || response.usageMetadata.promptTokenCount || 0,
            output: response.usageMetadata.candidatesTokens || response.usageMetadata.candidatesTokenCount || 0,
          },
          error: null,
        };
      }

      // Alternative format (OpenAI-compatible)
      if (response.usage) {
        return {
          data: {
            input: response.usage.prompt_tokens || response.usage.promptTokens || 0,
            output: response.usage.completion_tokens || response.usage.completionTokens || 0,
          },
          error: null,
        };
      }

      return {
        data: null,
        error: "No usage data found in Google AI response",
      };
    } catch (error) {
      return {
        data: null,
        error: `Failed to parse Google AI usage: ${error}`,
      };
    }
  }
}