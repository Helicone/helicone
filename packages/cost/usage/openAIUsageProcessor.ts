import { IUsageProcessor, ParseInput } from "./IUsageProcessor";
import { ModelUsage } from "./types";
import { Result } from "../../common/result";

export class OpenAIUsageProcessor implements IUsageProcessor {
  public async parse(parseInput: ParseInput): Promise<Result<ModelUsage, string>> {
    try {
      if (parseInput.isStream) {
        return this.parseStreamResponse(parseInput.responseBody);
      } else {
        return this.parseNonStreamResponse(parseInput.responseBody);
      }
    } catch (error) {
      return {
        data: null,
        error: `Failed to parse OpenAI usage: ${error}`,
      };
    }
  }

  private parseNonStreamResponse(responseBody: string): Result<ModelUsage, string> {
    try {
      const parsedResponse = JSON.parse(responseBody);
      const usage = this.extractUsageFromResponse(parsedResponse);
      
      return {
        data: usage,
        error: null,
      };
    } catch (error) {
      return {
        data: null,
        error: `Failed to parse non-stream response: ${error}`,
      };
    }
  }

  private parseStreamResponse(responseBody: string): Result<ModelUsage, string> {
    try {
      const lines = responseBody
        .split("\n")
        .filter((line) => line.trim() !== "" && !line.includes("OPENROUTER PROCESSING"))
        .map((line) => {
          if (line === "data: [DONE]") return null;
          try {
            return JSON.parse(line.replace("data: ", ""));
          } catch {
            return null;
          }
        })
        .filter((data) => data !== null);

      const consolidatedData = this.consolidateStreamData(lines);
      const usage = this.extractUsageFromResponse(consolidatedData);

      return {
        data: usage,
        error: null,
      };
    } catch (error) {
      return {
        data: null,
        error: `Failed to parse stream response: ${error}`,
      };
    }
  }

  private consolidateStreamData(streamData: any[]): any {
    const lastChunkWithUsage = [...streamData].reverse().find(chunk => chunk?.usage);
    if (lastChunkWithUsage?.usage) {
      return lastChunkWithUsage;
    }

    const consolidated: any = {
      choices: [],
      usage: null,
    };

    for (const chunk of streamData) {
      if (chunk?.usage) {
        consolidated.usage = chunk.usage;
      }
      if (chunk?.id) {
        consolidated.id = chunk.id;
      }
      if (chunk?.model) {
        consolidated.model = chunk.model;
      }
    }

    return consolidated;
  }

  private extractUsageFromResponse(parsedResponse: any): ModelUsage {
    if (!parsedResponse || typeof parsedResponse !== "object") {
      return {
        input: 0,
        output: 0,
      };
    }

    const usage = parsedResponse.usage || {};
    
    const promptTokens = usage.prompt_tokens ?? usage.input_tokens ?? 0;
    const completionTokens = usage.completion_tokens ?? usage.output_tokens ?? 0;
    
    const promptDetails = usage.prompt_tokens_details || {};
    const completionDetails = usage.completion_tokens_details || {};
    
    const cachedTokens = promptDetails.cached_tokens ?? 0;
    const promptAudioTokens = promptDetails.audio_tokens ?? 0;
    const completionAudioTokens = completionDetails.audio_tokens ?? 0;
    const reasoningTokens = completionDetails.reasoning_tokens ?? 0;
    
    const effectivePromptTokens = Math.max(0, promptTokens - cachedTokens - promptAudioTokens);
    const effectiveCompletionTokens = Math.max(0, completionTokens - completionAudioTokens - reasoningTokens);
    
    const modelUsage: ModelUsage = {
      input: effectivePromptTokens,
      output: effectiveCompletionTokens,
    };

    if (cachedTokens > 0) {
      modelUsage.cacheDetails = {
        cachedInput: cachedTokens,
      };
    }

    if (reasoningTokens > 0) {
      modelUsage.thinking = reasoningTokens;
    }

    if (promptAudioTokens > 0 || completionAudioTokens > 0) {
      // TODO: add audio output support since some models support it in the 
      // chat completions endpoint
      modelUsage.audio = promptAudioTokens + completionAudioTokens;
    }

    const rejectedTokens = completionDetails.rejected_prediction_tokens ?? 0;
    const acceptedTokens = completionDetails.accepted_prediction_tokens ?? 0;
    if (rejectedTokens > 0 || acceptedTokens > 0) {
      modelUsage.output = effectiveCompletionTokens + acceptedTokens;
    }

    return modelUsage;
  }
}