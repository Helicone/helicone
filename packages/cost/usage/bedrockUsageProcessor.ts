import { ModelUsage } from "./types";
import { AnthropicUsageProcessor } from "./anthropicUsageProcessor";

export class BedrockUsageProcessor extends AnthropicUsageProcessor {
  protected extractUsageFromResponse(parsedResponse: any): ModelUsage {
    if (!parsedResponse || typeof parsedResponse !== "object") {
      return {
        input: 0,
        output: 0,
      };
    }

    const usage = parsedResponse.usage || {};
    
    const inputTokens = usage.input_tokens ?? 0;
    const outputTokens = usage.output_tokens ?? 0;
    const cacheReadInputTokens = usage.cache_read_input_tokens ?? 0;

    // Bedrock does not support complex cache writes like Anthropic
    // and cache creation details are not available.

    const cacheWriteInputTokens = usage.cache_creation_input_tokens ?? 0;
    
    const modelUsage: ModelUsage = {
      input: inputTokens,
      output: outputTokens,
    };

    if (cacheReadInputTokens > 0 || cacheWriteInputTokens > 0) {
      modelUsage.cacheDetails = {
        cachedInput: cacheReadInputTokens,
        write5m: cacheWriteInputTokens,
      };
    }

    return modelUsage;
  }
}
