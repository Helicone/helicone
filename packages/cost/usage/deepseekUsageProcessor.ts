import { OpenAIUsageProcessor } from "./openAIUsageProcessor";
import { ModelUsage } from "./types";

export class DeepSeekUsageProcessor extends OpenAIUsageProcessor {
  protected extractUsageFromResponse(parsedResponse: any): ModelUsage {
    if (!parsedResponse || typeof parsedResponse !== "object") {
      return {
        input: 0,
        output: 0,
      };
    }

    const usage = parsedResponse.usage || {};

    // DeepSeek uses different field names for cache tracking
    const promptCacheHitTokens = usage.prompt_cache_hit_tokens ?? 0;
    const promptCacheMissTokens = usage.prompt_cache_miss_tokens ?? 0;
    const completionTokens =
      usage.completion_tokens ?? usage.output_tokens ?? 0;

    // Extract reasoning tokens if present (for deepseek-reasoner)
    const completionDetails = usage.completion_tokens_details || {};
    const reasoningTokens = completionDetails.reasoning_tokens ?? 0;

    // For DeepSeek:
    // - prompt_tokens = prompt_cache_hit_tokens + prompt_cache_miss_tokens
    // - The non-cached input is prompt_cache_miss_tokens
    // - The cached input is prompt_cache_hit_tokens
    // - completion_tokens includes reasoning_tokens, so we subtract them to get actual output
    const effectivePromptTokens = promptCacheMissTokens;
    const effectiveCompletionTokens = Math.max(
      0,
      completionTokens - reasoningTokens
    );

    const modelUsage: ModelUsage = {
      input: effectivePromptTokens,
      output: effectiveCompletionTokens,
    };

    // Add cache details if there are cached tokens
    if (promptCacheHitTokens > 0) {
      modelUsage.cacheDetails = {
        cachedInput: promptCacheHitTokens,
      };
    }

    // Add reasoning/thinking tokens if present
    if (reasoningTokens > 0) {
      modelUsage.thinking = reasoningTokens;
    }

    return modelUsage;
  }
}
