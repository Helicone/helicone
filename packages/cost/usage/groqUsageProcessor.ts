import { OpenAIUsageProcessor } from "./openAIUsageProcessor";

export class GroqUsageProcessor extends OpenAIUsageProcessor {
  // Groq puts usage in x_groq.usage in streaming responses
  protected consolidateStreamData(streamData: any[]): any {
    // Look for usage in x_groq.usage first (Groq's location)
    const lastChunkWithUsage = [...streamData].reverse().find(
      chunk => chunk?.usage || chunk?.x_groq?.usage
    );

    if (lastChunkWithUsage?.usage) {
      return lastChunkWithUsage;
    }

    // Handle Groq's x_groq.usage format
    if (lastChunkWithUsage?.x_groq?.usage) {
      return {
        ...lastChunkWithUsage,
        usage: lastChunkWithUsage.x_groq.usage
      };
    }

    // Fallback to building consolidated data
    const consolidated: any = {
      choices: [],
      usage: null,
    };

    for (const chunk of streamData) {
      if (chunk?.usage) {
        consolidated.usage = chunk.usage;
      } else if (chunk?.x_groq?.usage) {
        consolidated.usage = chunk.x_groq.usage;
      }
      if (chunk?.id) {
        consolidated.id = chunk.id;
      }
      if (chunk?.model) {
        consolidated.model = chunk.model;
      }
    }

    return consolidated;
  }
}