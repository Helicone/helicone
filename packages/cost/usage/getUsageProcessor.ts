import { OpenAIUsageProcessor } from "./openAIUsageProcessor";
import { AnthropicUsageProcessor } from "./anthropicUsageProcessor";
import { GroqUsageProcessor } from "./groqUsageProcessor";
import { XAIUsageProcessor } from "./xaiUsageProcessor";
import { OpenRouterUsageProcessor } from "./openRouterUsageProcessor";
import { DeepSeekUsageProcessor } from "./deepseekUsageProcessor";
import { VertexUsageProcessor } from "./vertexUsageProcessor";
import { GoogleUsageProcessor } from "./googleUsageProcessor";
import { BedrockUsageProcessor } from "./bedrockUsageProcessor";
import { IUsageProcessor } from "./IUsageProcessor";
import { ModelProviderName } from "../models/providers";

export function getUsageProcessor(
  provider: ModelProviderName
): IUsageProcessor | null {
  switch (provider) {
    case "openai":
    case "azure":
    case "deepinfra":
    case "novita":
      return new OpenAIUsageProcessor();
    case "anthropic":
      return new AnthropicUsageProcessor();
    case "groq":
      return new GroqUsageProcessor();
    case "xai":
      return new XAIUsageProcessor();
    case "openrouter":
      return new OpenRouterUsageProcessor();
    case "deepseek":
      return new DeepSeekUsageProcessor();
    case "vertex":
      return new VertexUsageProcessor();
    case "google-ai-studio":
      return new GoogleUsageProcessor();
    case "bedrock":
      return new BedrockUsageProcessor();
    default:
      return null;
  }
}
