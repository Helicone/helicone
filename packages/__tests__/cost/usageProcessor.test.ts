import { describe, it, expect } from "@jest/globals";
import { OpenAIUsageProcessor } from "@helicone-package/cost/usage/openAIUsageProcessor";
import { getUsageProcessor } from "@helicone-package/cost/usage/getUsageProcessor";
import * as fs from "fs";
import * as path from "path";

describe("getUsageProcessor", () => {
  it("should return OpenAIUsageProcessor for openai provider", () => {
    const processor = getUsageProcessor("openai");
    expect(processor).toBeInstanceOf(OpenAIUsageProcessor);
  });

  it("should throw error for unsupported provider", () => {
    expect(() => {
      getUsageProcessor("unsupported-provider" as any);
    }).toThrow("Usage processor not found for provider: unsupported-provider");
  });
});

describe("OpenAIUsageProcessor", () => {
  const processor = new OpenAIUsageProcessor();

  it("should parse real GPT-4o response with cached tokens", async () => {
    const responseData = fs.readFileSync(
      path.join(__dirname, "testData", "gpt4o-response-cached.txt"),
      "utf-8"
    );

    const result = await processor.parse({ 
      responseBody: responseData, 
      isStream: false 
    });

    expect(result.error).toBeNull();
    expect(result.data).toEqual({
      input: 96,
      output: 10,
      cacheDetails: {
        cachedInput: 1152
      }
    });
  });

  it("should parse real GPT-4o stream response", async () => {
    const streamData = fs.readFileSync(
      path.join(__dirname, "testData", "gpt4o-stream-response.txt"),
      "utf-8"
    );

    const result = await processor.parse({ 
      responseBody: streamData, 
      isStream: true 
    });

    expect(result.error).toBeNull();
    expect(result.data).toEqual({
      input: 1248,
      output: 10
    });
  });

  it("usage processing snapshot", async () => {
    const testCases = [
      {
        name: "cached-response",
        data: fs.readFileSync(path.join(__dirname, "testData", "gpt4o-response-cached.txt"), "utf-8"),
        isStream: false
      },
      {
        name: "stream-response", 
        data: fs.readFileSync(path.join(__dirname, "testData", "gpt4o-stream-response.txt"), "utf-8"),
        isStream: true
      }
    ];

    const results: Record<string, any> = {};
    
    for (const testCase of testCases) {
      const result = await processor.parse({
        responseBody: testCase.data,
        isStream: testCase.isStream
      });
      results[testCase.name] = result;
    }

    expect(results).toMatchSnapshot();
  });
}); 