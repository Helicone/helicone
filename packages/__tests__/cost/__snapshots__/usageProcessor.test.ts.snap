// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenAIUsageProcessor usage processing snapshot 1`] = `
{
  "cached-response": {
    "data": {
      "cacheDetails": {
        "cachedInput": 1152,
      },
      "input": 96,
      "output": 10,
    },
    "error": null,
  },
  "stream-response": {
    "data": {
      "input": 1248,
      "output": 10,
    },
    "error": null,
  },
}
`;
